File: android/src/main/java/io/getstream/webrtc/flutter/FlutterRTCEncryptionManager.java
package io.getstream.webrtc.flutter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.webrtc.EncryptionManager;
import org.webrtc.RtpReceiver;
import org.webrtc.RtpSender;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.Result;
import io.getstream.webrtc.flutter.utils.AnyThreadSink;

/** Bridges {@link EncryptionManager} to Dart using Stream's implementation via a per-instance event channel. */
class FlutterRTCEncryptionManager {

  /** Dart sends -1 when it wants native to infer audio vs video from RTP. */
  private static final int TRACK_TYPE_UNSPECIFIED = -1;

  private final StateProvider stateProvider;
  private final ConcurrentHashMap<String, Handle> handles = new ConcurrentHashMap<>();

  /**
   * Releases native managers off the platform thread.
   *
   * <p>Disposing joins the manager's frame-crypto worker, so doing it inline
   * would block the Android main thread for as long as that worker takes to
   * drain — an ANR if it is mid-frame or wedged. Nothing after teardown needs
   * the manager, so the join can finish in the background.
   */
  private final ExecutorService disposeExecutor =
      Executors.newSingleThreadExecutor(
          runnable -> {
            Thread thread = new Thread(runnable, "E2EEManagerDispose");
            thread.setDaemon(true);
            return thread;
          });

  FlutterRTCEncryptionManager(StateProvider stateProvider) {
    this.stateProvider = stateProvider;
  }

  private static class Handle implements EventChannel.StreamHandler {
    final EncryptionManager manager;
    final EventChannel eventChannel;
    @Nullable EventChannel.EventSink sink;

    Handle(EncryptionManager manager, EventChannel eventChannel) {
      this.manager = manager;
      this.eventChannel = eventChannel;
    }

    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {
      sink = new AnyThreadSink(events);
    }

    @Override
    public void onCancel(Object arguments) {
      sink = null;
    }

    void send(Map<String, Object> event) {
      final EventChannel.EventSink target = sink;
      if (target != null) {
        target.success(event);
      }
    }

    /**
     * Detaches the event channel. Must run on the platform thread, and must
     * happen before {@link #releaseNative()} so no event races the teardown.
     */
    void detach() {
      eventChannel.setStreamHandler(null);
      sink = null;
      manager.setObserver(null);
    }

    /** Releases the native manager. Blocks on its frame-crypto worker. */
    void releaseNative() {
      manager.dispose();
    }
  }

  /**
   * @return {@code true} when {@code call} was an encryption-manager method,
   *     handled (successfully or with an error) by this class.
   */
  boolean handleMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    final String method = call.method;
    if (method == null || !method.startsWith("encryptionManager")) {
      return false;
    }

    switch (method) {
      case "encryptionManagerCreate":
        create(call, result);
        return true;
      case "encryptionManagerSetKey":
        setKey(call, result);
        return true;
      case "encryptionManagerSetSharedKey":
        setSharedKey(call, result);
        return true;
      case "encryptionManagerRemoveKey":
        removeKey(call, result);
        return true;
      case "encryptionManagerRemoveAllKeys":
        removeAllKeys(call, result);
        return true;
      case "encryptionManagerRemoveSharedKey":
        removeSharedKey(call, result);
        return true;
      case "encryptionManagerEncrypt":
        encrypt(call, result);
        return true;
      case "encryptionManagerDecrypt":
        decrypt(call, result);
        return true;
      case "encryptionManagerEnablePerformanceReporting":
        enablePerformanceReporting(call, result);
        return true;
      case "encryptionManagerRequestKeyState":
        requestKeyState(call, result);
        return true;
      case "encryptionManagerDispose":
        dispose(call, result);
        return true;
      default:
        return false;
    }
  }

  /** Releases every manager, e.g. when the plugin detaches from the engine. */
  void disposeAll() {
    final List<Handle> pending = new ArrayList<>(handles.values());
    handles.clear();
    for (Handle handle : pending) {
      handle.detach();
      disposeExecutor.execute(handle::releaseNative);
    }
  }

  // MARK: - Methods

  private void create(MethodCall call, Result result) {
    final String userId = call.argument("userId");
    if (userId == null || userId.isEmpty()) {
      result.error("encryptionManagerCreateFailed", "userId is required", null);
      return;
    }

    final Integer algorithmValue = call.argument("algorithm");
    final EncryptionManager.Algorithm algorithm =
        algorithmValue != null && algorithmValue == EncryptionManager.Algorithm.AES_256_GCM.getValue()
            ? EncryptionManager.Algorithm.AES_256_GCM
            : EncryptionManager.Algorithm.AES_128_GCM;

    try {
      final String managerId = UUID.randomUUID().toString();
      final EncryptionManager manager = EncryptionManager.create(userId, algorithm);
      final EventChannel eventChannel =
          new EventChannel(stateProvider.getMessenger(), "FlutterWebRTC/e2ee/" + managerId);
      final Handle handle = new Handle(manager, eventChannel);

      eventChannel.setStreamHandler(handle);
      manager.setObserver(event -> handle.send(eventToMap(event)));
      handles.put(managerId, handle);

      final Map<String, Object> response = new HashMap<>();
      response.put("managerId", managerId);
      result.success(response);
    } catch (Exception e) {
      result.error("encryptionManagerCreateFailed", e.getMessage(), null);
    }
  }

  private void setKey(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    final String userId = call.argument("userId");
    final Integer keyIndex = call.argument("keyIndex");
    final byte[] rawKey = call.argument("rawKey");
    if (userId == null || keyIndex == null || rawKey == null) {
      result.error(call.method + "Failed", "userId, keyIndex and rawKey are required", null);
      return;
    }
    try {
      manager.setKey(userId, keyIndex, rawKey);
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void setSharedKey(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    final Integer keyIndex = call.argument("keyIndex");
    final byte[] rawKey = call.argument("rawKey");
    if (keyIndex == null || rawKey == null) {
      result.error(call.method + "Failed", "keyIndex and rawKey are required", null);
      return;
    }
    try {
      manager.setSharedKey(keyIndex, rawKey);
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void removeKey(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    final String userId = call.argument("userId");
    final Integer keyIndex = call.argument("keyIndex");
    if (userId == null || keyIndex == null) {
      result.error(call.method + "Failed", "userId and keyIndex are required", null);
      return;
    }
    try {
      manager.removeKey(userId, keyIndex);
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void removeAllKeys(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    final String userId = call.argument("userId");
    if (userId == null) {
      result.error(call.method + "Failed", "userId is required", null);
      return;
    }
    try {
      manager.removeAllKeys(userId);
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void removeSharedKey(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    final Integer keyIndex = call.argument("keyIndex");
    if (keyIndex == null) {
      result.error(call.method + "Failed", "keyIndex is required", null);
      return;
    }
    try {
      manager.removeSharedKey(keyIndex);
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void encrypt(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }

    final PeerConnectionObserver pco = requireObserver(call, result);
    if (pco == null) {
      return;
    }

    final String senderId = call.argument("rtpSenderId");
    if (senderId == null) {
      result.error(call.method + "Failed", "rtpSenderId is required", null);
      return;
    }

    final RtpSender sender = pco.getRtpSenderById(senderId);
    if (sender == null) {
      result.error(call.method + "Failed", "sender " + senderId + " not found", null);
      return;
    }

    try {
      manager.encrypt(sender, call.argument("codec"), trackType(call));
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void decrypt(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }

    final PeerConnectionObserver pco = requireObserver(call, result);
    if (pco == null) {
      return;
    }

    final String receiverId = call.argument("rtpReceiverId");
    final String userId = call.argument("userId");
    if (receiverId == null || userId == null || userId.isEmpty()) {
      result.error(call.method + "Failed", "rtpReceiverId and userId are required", null);
      return;
    }

    final RtpReceiver receiver = pco.getRtpReceiverById(receiverId);
    if (receiver == null) {
      result.error(call.method + "Failed", "receiver " + receiverId + " not found", null);
      return;
    }

    try {
      manager.decrypt(receiver, userId, trackType(call));
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void enablePerformanceReporting(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    final Boolean enabled = call.argument("enabled");
    try {
      manager.enablePerformanceReporting(Boolean.TRUE.equals(enabled));
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void requestKeyState(MethodCall call, Result result) {
    final EncryptionManager manager = requireManager(call, result);
    if (manager == null) {
      return;
    }
    try {
      manager.requestKeyState();
      result.success(null);
    } catch (Exception e) {
      result.error(call.method + "Failed", e.getMessage(), null);
    }
  }

  private void dispose(MethodCall call, Result result) {
    final String managerId = call.argument("managerId");
    final Handle handle = managerId == null ? null : handles.remove(managerId);
    if (handle != null) {
      handle.detach();
      // Off the platform thread: releasing joins the frame-crypto worker.
      disposeExecutor.execute(handle::releaseNative);
    }
    // Disposing an unknown manager is not an error: Dart may retry teardown.
    // Answering before the join completes is fine — the handle is already out
    // of the registry, so nothing can reach it again.
    result.success(null);
  }

  // MARK: - Helpers

  @Nullable
  private EncryptionManager requireManager(MethodCall call, Result result) {
    final String managerId = call.argument("managerId");
    final Handle handle = managerId == null ? null : handles.get(managerId);
    if (handle == null) {
      result.error(
          call.method + "Failed", "EncryptionManager " + managerId + " not found", null);
      return null;
    }
    return handle.manager;
  }

  @Nullable
  private PeerConnectionObserver requireObserver(MethodCall call, Result result) {
    final String peerConnectionId = call.argument("peerConnectionId");
    final PeerConnectionObserver pco =
        peerConnectionId == null ? null : stateProvider.getPeerConnectionObserver(peerConnectionId);
    if (pco == null) {
      result.error(
          call.method + "Failed", "peerConnection " + peerConnectionId + " not found", null);
      return null;
    }
    return pco;
  }

  /** Maps Dart's {@code trackType} to the enum, or {@code null} to let RTP decide. */
  @Nullable
  private static EncryptionManager.TrackType trackType(MethodCall call) {
    final Integer value = call.argument("trackType");
    if (value == null || value == TRACK_TYPE_UNSPECIFIED) {
      return null;
    }
    for (EncryptionManager.TrackType type : EncryptionManager.TrackType.values()) {
      if (type.getValue() == value) {
        return type;
      }
    }
    return null;
  }

  // MARK: - Event serialization

  private static Map<String, Object> eventToMap(EncryptionManager.E2eeEvent event) {
    final Map<String, Object> map = new HashMap<>();
    map.put("type", event.type.getValue());
    map.put("name", event.name);
    map.put("userId", event.userId);
    if (event.trackType != null) {
      map.put("trackType", event.trackType.getValue());
    }
    if (event.keyIndex != null) {
      map.put("keyIndex", event.keyIndex);
    }
    if (event.version != null) {
      map.put("version", event.version);
    }
    if (event.reason != null) {
      map.put("reason", event.reason);
    }
    if (event.keyState != null) {
      map.put("keyState", keyStateToMap(event.keyState));
    }
    if (event.encode != null) {
      map.put("encode", perfToList(event.encode));
    }
    if (event.decode != null) {
      map.put("decode", perfToList(event.decode));
    }
    return map;
  }

  private static Map<String, Object> keyStateToMap(EncryptionManager.KeyStateReport report) {
    final List<Object> perUserKeys = new ArrayList<>();
    for (EncryptionManager.UserKey key : report.perUserKeys) {
      final Map<String, Object> entry = new HashMap<>();
      entry.put("userId", key.userId);
      entry.put("keyIndex", key.keyIndex);
      entry.put("fingerprint", key.fingerprint);
      perUserKeys.add(entry);
    }

    final List<Object> sharedKeys = new ArrayList<>();
    for (EncryptionManager.SharedKey key : report.sharedKeys) {
      final Map<String, Object> entry = new HashMap<>();
      entry.put("keyIndex", key.keyIndex);
      entry.put("fingerprint", key.fingerprint);
      entry.put("isActive", key.isActive);
      sharedKeys.add(entry);
    }

    final Map<String, Object> map = new HashMap<>();
    map.put("perUserKeys", perUserKeys);
    map.put("sharedKeys", sharedKeys);
    return map;
  }

  private static List<Object> perfToList(List<EncryptionManager.TrackPerf> samples) {
    final List<Object> list = new ArrayList<>();
    for (EncryptionManager.TrackPerf sample : samples) {
      final Map<String, Object> entry = new HashMap<>();
      entry.put("userId", sample.userId);
      if (sample.trackType != null) {
        entry.put("trackType", sample.trackType.getValue());
      }
      if (sample.codec != null) {
        entry.put("codec", sample.codec);
      }
      entry.put("fps", sample.fps);
      entry.put("maxCryptoMs", sample.maxCryptoMs);
      list.add(entry);
    }
    return list;
  }
}

File: ios/stream_webrtc_flutter/Sources/stream_webrtc_flutter/FlutterRTCEncryptionManager.m
#import "include/stream_webrtc_flutter/FlutterRTCEncryptionManager.h"

/** Dart sends -1 when it wants native to infer audio vs video from RTP. */
static const NSInteger kTrackTypeUnspecified = -1;

/**
 * Manager registry.
 *
 * The plugin is a singleton (`+[FlutterWebRTCPlugin sharedSingleton]`) and a
 * category cannot add storage, so the registry lives here. Guarded by
 * @synchronized because Dart calls arrive on the platform thread while
 * `dispose` can run from teardown.
 */
static NSMutableDictionary<NSString*, FlutterRTCEncryptionManagerHandle*>* gHandles;

static NSMutableDictionary<NSString*, FlutterRTCEncryptionManagerHandle*>* handleRegistry(void) {
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    gHandles = [NSMutableDictionary dictionary];
  });
  return gHandles;
}

#pragma mark - Argument coercion

/**
 * Method-channel arguments are untrusted: Dart's `null` arrives as `NSNull`,
 * which raises on `integerValue`/`boolValue`. Returns nil unless the value is
 * a real number, so callers can reject or default without converting first.
 */
static NSNumber* _Nullable numberArg(id args, NSString* key) {
  if (![args isKindOfClass:[NSDictionary class]]) {
    return nil;
  }
  id value = ((NSDictionary*)args)[key];
  return [value isKindOfClass:[NSNumber class]] ? (NSNumber*)value : nil;
}

#pragma mark - Event serialization

static NSDictionary* userKeyToMap(RTCEncryptionUserKey* key) {
  return @{
    @"userId" : key.userId ?: @"",
    @"keyIndex" : @(key.keyIndex),
    @"fingerprint" : key.fingerprint ?: @""
  };
}

static NSDictionary* sharedKeyToMap(RTCEncryptionSharedKey* key) {
  return @{
    @"keyIndex" : @(key.keyIndex),
    @"fingerprint" : key.fingerprint ?: @"",
    @"isActive" : @(key.isActive)
  };
}

static NSDictionary* keyStateToMap(RTCEncryptionKeyState* keyState) {
  NSMutableArray* perUserKeys = [NSMutableArray array];
  for (RTCEncryptionUserKey* key in keyState.perUserKeys) {
    [perUserKeys addObject:userKeyToMap(key)];
  }

  NSMutableArray* sharedKeys = [NSMutableArray array];
  for (RTCEncryptionSharedKey* key in keyState.sharedKeys) {
    [sharedKeys addObject:sharedKeyToMap(key)];
  }

  return @{@"perUserKeys" : perUserKeys, @"sharedKeys" : sharedKeys};
}

static NSArray* perfToList(NSArray<RTCEncryptionTrackPerf*>* samples) {
  NSMutableArray* list = [NSMutableArray array];
  for (RTCEncryptionTrackPerf* sample in samples) {
    NSMutableDictionary* entry = [NSMutableDictionary dictionary];
    entry[@"userId"] = sample.userId ?: @"";
    entry[@"trackType"] = @(sample.trackType);
    if (sample.codec != nil) {
      entry[@"codec"] = sample.codec;
    }
    entry[@"fps"] = @(sample.fps);
    entry[@"maxCryptoMs"] = @(sample.maxCryptoMs);
    [list addObject:entry];
  }
  return list;
}

static NSDictionary* eventToMap(RTCE2eeEvent* event) {
  NSMutableDictionary* map = [NSMutableDictionary dictionary];
  map[@"type"] = @(event.type);
  map[@"name"] = event.name ?: @"";
  map[@"userId"] = event.userId ?: @"";
  if (event.trackType != nil) {
    map[@"trackType"] = event.trackType;
  }
  if (event.keyIndex != nil) {
    map[@"keyIndex"] = event.keyIndex;
  }
  if (event.version != nil) {
    map[@"version"] = event.version;
  }
  if (event.reason != nil) {
    map[@"reason"] = event.reason;
  }
  if (event.keyState != nil) {
    map[@"keyState"] = keyStateToMap(event.keyState);
  }
  if (event.encode != nil) {
    map[@"encode"] = perfToList(event.encode);
  }
  if (event.decode != nil) {
    map[@"decode"] = perfToList(event.decode);
  }
  return map;
}

#pragma mark - Handle

@implementation FlutterRTCEncryptionManagerHandle {
  FlutterEventSink _eventSink;
}

- (instancetype)initWithManager:(RTCEncryptionManager*)manager
                   eventChannel:(FlutterEventChannel*)eventChannel {
  self = [super init];
  if (self) {
    _manager = manager;
    _eventChannel = eventChannel;
  }
  return self;
}

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events {
  _eventSink = events;
  return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
  _eventSink = nil;
  return nil;
}

- (void)encryptionManager:(RTCEncryptionManager*)manager didReceiveEvent:(RTCE2eeEvent*)event {
  postEvent(_eventSink, eventToMap(event));
}

- (void)detach {
  // Dropping the delegate first stops events racing the teardown.
  _manager.delegate = nil;
  [_eventChannel setStreamHandler:nil];
  _eventSink = nil;
}

- (void)releaseNative {
  [_manager dispose];
}

@end

#pragma mark - Plugin category

@implementation FlutterWebRTCPlugin (EncryptionManager)

- (BOOL)handleEncryptionManagerMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* method = call.method;
  if (![method hasPrefix:@"encryptionManager"]) {
    return NO;
  }

  if ([@"encryptionManagerCreate" isEqualToString:method]) {
    [self encryptionManagerCreate:call result:result];
  } else if ([@"encryptionManagerSetKey" isEqualToString:method]) {
    [self encryptionManagerSetKey:call result:result];
  } else if ([@"encryptionManagerSetSharedKey" isEqualToString:method]) {
    [self encryptionManagerSetSharedKey:call result:result];
  } else if ([@"encryptionManagerRemoveKey" isEqualToString:method]) {
    [self encryptionManagerRemoveKey:call result:result];
  } else if ([@"encryptionManagerRemoveAllKeys" isEqualToString:method]) {
    [self encryptionManagerRemoveAllKeys:call result:result];
  } else if ([@"encryptionManagerRemoveSharedKey" isEqualToString:method]) {
    [self encryptionManagerRemoveSharedKey:call result:result];
  } else if ([@"encryptionManagerEncrypt" isEqualToString:method]) {
    [self encryptionManagerEncrypt:call result:result];
  } else if ([@"encryptionManagerDecrypt" isEqualToString:method]) {
    [self encryptionManagerDecrypt:call result:result];
  } else if ([@"encryptionManagerEnablePerformanceReporting" isEqualToString:method]) {
    [self encryptionManagerEnablePerformanceReporting:call result:result];
  } else if ([@"encryptionManagerRequestKeyState" isEqualToString:method]) {
    [self encryptionManagerRequestKeyState:call result:result];
  } else if ([@"encryptionManagerDispose" isEqualToString:method]) {
    [self encryptionManagerDispose:call result:result];
  } else {
    return NO;
  }

  return YES;
}

- (void)disposeAllEncryptionManagers {
  NSArray<FlutterRTCEncryptionManagerHandle*>* handles;
  @synchronized(handleRegistry()) {
    handles = handleRegistry().allValues;
    [handleRegistry() removeAllObjects];
  }
  for (FlutterRTCEncryptionManagerHandle* handle in handles) {
    [handle detach];
  }
  // Releasing joins the manager's frame-crypto worker, which would block the
  // platform thread for as long as that worker takes to drain.
  dispatch_async(dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
    for (FlutterRTCEncryptionManagerHandle* handle in handles) {
      [handle releaseNative];
    }
  });
}

#pragma mark - Methods

- (void)encryptionManagerCreate:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSDictionary* args = call.arguments;
  NSString* userId = args[@"userId"];
  if (![userId isKindOfClass:[NSString class]] || userId.length == 0) {
    result([FlutterError errorWithCode:@"encryptionManagerCreateFailed"
                               message:@"userId is required"
                               details:nil]);
    return;
  }

  // A missing algorithm keeps the AES-128 default; a malformed one is rejected
  // rather than silently downgraded.
  id algorithmValue = [args isKindOfClass:[NSDictionary class]] ? args[@"algorithm"] : nil;
  if (algorithmValue != nil && ![algorithmValue isKindOfClass:[NSNumber class]]) {
    [self failCall:call result:result message:@"algorithm must be a number"];
    return;
  }
  RTCEncryptionAlgorithm algorithm =
      [algorithmValue integerValue] == RTCEncryptionAlgorithmAes256Gcm
          ? RTCEncryptionAlgorithmAes256Gcm
          : RTCEncryptionAlgorithmAes128Gcm;

  NSError* error = nil;
  RTCEncryptionManager* manager = [RTCEncryptionManager createWithUserId:userId
                                                               algorithm:algorithm
                                                                   error:&error];
  if (manager == nil) {
    result([FlutterError errorWithCode:@"encryptionManagerCreateFailed"
                               message:error.localizedDescription ?: @"create failed"
                               details:nil]);
    return;
  }

  NSString* managerId = [[NSUUID UUID] UUIDString];
  FlutterEventChannel* eventChannel = [FlutterEventChannel
      eventChannelWithName:[NSString stringWithFormat:@"FlutterWebRTC/e2ee/%@", managerId]
           binaryMessenger:self.messenger];

  FlutterRTCEncryptionManagerHandle* handle =
      [[FlutterRTCEncryptionManagerHandle alloc] initWithManager:manager eventChannel:eventChannel];
  [eventChannel setStreamHandler:handle];
  manager.delegate = handle;

  @synchronized(handleRegistry()) {
    handleRegistry()[managerId] = handle;
  }

  result(@{@"managerId" : managerId});
}

- (void)encryptionManagerSetKey:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSDictionary* args = call.arguments;
  NSString* userId = args[@"userId"];
  NSNumber* keyIndex = numberArg(args, @"keyIndex");
  FlutterStandardTypedData* rawKey = args[@"rawKey"];
  if (userId == nil || keyIndex == nil || rawKey == nil) {
    [self failCall:call result:result message:@"userId, keyIndex and rawKey are required"];
    return;
  }

  NSError* error = nil;
  if (![manager setKey:userId keyIndex:keyIndex.intValue rawKey:rawKey.data error:&error]) {
    [self failCall:call result:result message:error.localizedDescription ?: @"setKey failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerSetSharedKey:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSDictionary* args = call.arguments;
  NSNumber* keyIndex = numberArg(args, @"keyIndex");
  FlutterStandardTypedData* rawKey = args[@"rawKey"];
  if (keyIndex == nil || rawKey == nil) {
    [self failCall:call result:result message:@"keyIndex and rawKey are required"];
    return;
  }

  NSError* error = nil;
  if (![manager setSharedKey:keyIndex.intValue rawKey:rawKey.data error:&error]) {
    [self failCall:call result:result message:error.localizedDescription ?: @"setSharedKey failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerRemoveKey:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSDictionary* args = call.arguments;
  NSString* userId = args[@"userId"];
  NSNumber* keyIndex = numberArg(args, @"keyIndex");
  if (userId == nil || keyIndex == nil) {
    [self failCall:call result:result message:@"userId and keyIndex are required"];
    return;
  }

  NSError* error = nil;
  if (![manager removeKey:userId keyIndex:keyIndex.intValue error:&error]) {
    [self failCall:call result:result message:error.localizedDescription ?: @"removeKey failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerRemoveAllKeys:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSString* userId = call.arguments[@"userId"];
  if (userId == nil) {
    [self failCall:call result:result message:@"userId is required"];
    return;
  }

  NSError* error = nil;
  if (![manager removeAllKeys:userId error:&error]) {
    [self failCall:call
            result:result
           message:error.localizedDescription ?: @"removeAllKeys failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerRemoveSharedKey:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSNumber* keyIndex = numberArg(call.arguments, @"keyIndex");
  if (keyIndex == nil) {
    [self failCall:call result:result message:@"keyIndex is required"];
    return;
  }

  NSError* error = nil;
  if (![manager removeSharedKey:keyIndex.intValue error:&error]) {
    [self failCall:call
            result:result
           message:error.localizedDescription ?: @"removeSharedKey failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerEncrypt:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  RTCPeerConnection* peerConnection = [self requirePeerConnectionForCall:call result:result];
  if (peerConnection == nil) {
    return;
  }

  NSString* senderId = call.arguments[@"rtpSenderId"];
  if (senderId == nil) {
    [self failCall:call result:result message:@"rtpSenderId is required"];
    return;
  }

  RTCRtpSender* sender = [self getRtpSenderById:peerConnection Id:senderId];
  if (sender == nil) {
    [self failCall:call
            result:result
           message:[NSString stringWithFormat:@"sender %@ not found", senderId]];
    return;
  }

  NSString* codec = call.arguments[@"codec"];
  if (![codec isKindOfClass:[NSString class]]) {
    codec = nil;
  }

  NSError* error = nil;
  if (![manager encrypt:sender codec:codec trackType:[self trackTypeForCall:call] error:&error]) {
    [self failCall:call result:result message:error.localizedDescription ?: @"encrypt failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerDecrypt:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  RTCPeerConnection* peerConnection = [self requirePeerConnectionForCall:call result:result];
  if (peerConnection == nil) {
    return;
  }

  NSDictionary* args = call.arguments;
  NSString* receiverId = args[@"rtpReceiverId"];
  NSString* userId = args[@"userId"];
  if (receiverId == nil || userId == nil || userId.length == 0) {
    [self failCall:call result:result message:@"rtpReceiverId and userId are required"];
    return;
  }

  RTCRtpReceiver* receiver = [self getRtpReceiverById:peerConnection Id:receiverId];
  if (receiver == nil) {
    [self failCall:call
            result:result
           message:[NSString stringWithFormat:@"receiver %@ not found", receiverId]];
    return;
  }

  NSError* error = nil;
  if (![manager decrypt:receiver
                 userId:userId
              trackType:[self trackTypeForCall:call]
                  error:&error]) {
    [self failCall:call result:result message:error.localizedDescription ?: @"decrypt failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerEnablePerformanceReporting:(FlutterMethodCall*)call
                                             result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSNumber* enabled = numberArg(call.arguments, @"enabled");
  if (enabled == nil) {
    [self failCall:call result:result message:@"enabled is required"];
    return;
  }

  NSError* error = nil;
  if (![manager enablePerformanceReporting:enabled.boolValue error:&error]) {
    [self failCall:call
            result:result
           message:error.localizedDescription ?: @"enablePerformanceReporting failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerRequestKeyState:(FlutterMethodCall*)call result:(FlutterResult)result {
  RTCEncryptionManager* manager = [self requireManagerForCall:call result:result];
  if (manager == nil) {
    return;
  }

  NSError* error = nil;
  if (![manager requestKeyState:&error]) {
    [self failCall:call
            result:result
           message:error.localizedDescription ?: @"requestKeyState failed"];
    return;
  }
  result(nil);
}

- (void)encryptionManagerDispose:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* managerId = call.arguments[@"managerId"];
  FlutterRTCEncryptionManagerHandle* handle = nil;
  if (managerId != nil) {
    @synchronized(handleRegistry()) {
      handle = handleRegistry()[managerId];
      [handleRegistry() removeObjectForKey:managerId];
    }
  }
  [handle detach];
  if (handle != nil) {
    // Off the platform thread: releasing joins the frame-crypto worker.
    // Answering first is fine — the handle is already out of the registry.
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_UTILITY, 0), ^{
      [handle releaseNative];
    });
  }
  // Disposing an unknown manager is not an error: Dart may retry teardown.
  result(nil);
}

#pragma mark - Helpers

- (nullable RTCEncryptionManager*)requireManagerForCall:(FlutterMethodCall*)call
                                                 result:(FlutterResult)result {
  NSString* managerId = call.arguments[@"managerId"];
  FlutterRTCEncryptionManagerHandle* handle = nil;
  if (managerId != nil) {
    @synchronized(handleRegistry()) {
      handle = handleRegistry()[managerId];
    }
  }
  if (handle == nil) {
    [self failCall:call
            result:result
           message:[NSString stringWithFormat:@"EncryptionManager %@ not found", managerId]];
    return nil;
  }
  return handle.manager;
}

- (nullable RTCPeerConnection*)requirePeerConnectionForCall:(FlutterMethodCall*)call
                                                     result:(FlutterResult)result {
  NSString* peerConnectionId = call.arguments[@"peerConnectionId"];
  RTCPeerConnection* peerConnection =
      peerConnectionId == nil ? nil : self.peerConnections[peerConnectionId];
  if (peerConnection == nil) {
    [self failCall:call
            result:result
           message:[NSString stringWithFormat:@"peerConnection %@ not found", peerConnectionId]];
    return nil;
  }
  return peerConnection;
}

/** Maps Dart's `trackType` to a boxed enum, or nil to let RTP decide. */
- (nullable NSNumber*)trackTypeForCall:(FlutterMethodCall*)call {
  NSNumber* value = numberArg(call.arguments, @"trackType");
  if (value == nil || value.integerValue == kTrackTypeUnspecified) {
    return nil;
  }
  return value;
}

- (void)failCall:(FlutterMethodCall*)call result:(FlutterResult)result message:(NSString*)message {
  result([FlutterError errorWithCode:[NSString stringWithFormat:@"%@Failed", call.method]
                             message:message
                             details:nil]);
}

@end
